### PR TITLE
Implement Plutarch Rationals

### PR DESCRIPTION
**Pull request by Geometer1729**
I'm not certain this is the best way to do Rationals in Plutarch, but we're going to need them fairly soon for Danaswap and as long as any future improvements can be backwards compatible with the Num and Fractional instances this should at least allow us to write and test code using rationals for now.

### CODE REVIEW

File: Plutarch/Rational.hs [hunk 1]
```diff
@@ -1,28 +1,118 @@
...
-    pinl (pfst (punRational x)) $ \x1 ->
-    pinl (psnd (punRational x)) $ \x2 ->
-    pinl (pfst (punRational y)) $ \y1 ->
-    pinl (psnd (punRational y)) $ \y2 ->
-    prational (x1 * y1) (x2 * y2)
-    -}
+module Plutarch.Rational 
+  (PRational
+  ,preduce
+  ,pnumerator
```
Comment (L-as) [inline Plutarch/Rational.hs:6]: ./bin/format

nix build .#check.x86_64-linux

File: Plutarch/Rational.hs [hunk 2]
```diff
@@ -1,28 +1,118 @@
...
+  x / y = preduce #
+    pmatch x (\(PRational xn xd) ->
+    pmatch y (\(PRational yn yd) ->
+      pcon (PRational (xn*yd) (xd*yn))
+             ))
+  fromRational r = 
+    pcon $ PRational (fromInteger $ numerator r) (fromInteger $ denominator r)
+
+preduce :: Term s (PRational :--> PRational)
+preduce = plam $ \x ->
```
Comment (L-as) [inline Plutarch/Rational.hs:88]: You need to hoist all your terms.

File: Plutarch/Rational.hs [hunk 3]
```diff
@@ -1,28 +1,118 @@
...
+        1
+  fromInteger n = pcon $ PRational (fromInteger n) 1
+            
+instance Fractional (Term s PRational) where
+  recip x = 
+    pmatch x (\(PRational xn xd) ->
+      pcon (PRational xd xn))
+  x / y = preduce #
+    pmatch x (\(PRational xn xd) ->
+    pmatch y (\(PRational yn yd) ->
```
Comment (L-as) [inline Plutarch/Rational.hs:81]: Please do `pmatch x $ ...`

File: Plutarch/Rational.hs [hunk 4]
```diff
@@ -1,28 +1,118 @@
...
+pgcd' = pfix #$ plam $ f
+  where
+    f self a' b' =
+        plet (pmax # a' # b') $ \a ->
+        plet (pmin # a' # b') $ \b ->
+          pif (b #== 0)
+            a
+            (self # (pmod # a # b) # b)
+
+pmin,pmax :: POrd a => Term s (a :--> a :--> a)
```
Comment (L-as) [inline Plutarch/Rational.hs:109]: Please don't declare multiple functions together. See github.com/mlabs-haskell/styleguide

**Review by L-as — CHANGES_REQUESTED** [review]
Waiting for the review of @kozross

**Comment by Geometer1729** [comment]
The arithmetic terms don't seem to be hoist-able because they are not term level functions. `pgcd'`,`pmin` and `pmax` are all used just once. For `pnumerator` , `pdenominator` and `pfromInteger` I'm not sure if I should hoist them. They are all pretty simple.

**Comment by srid** [comment]
I think we are also going to need a `PIsData` instance, to be able to store rationals in datum/redeemer.

**Comment by L-as** [comment]
> pgcd',pmin and pmax are all used just once. For pnumerator , pdenominator and pfromInteger I'm not sure if I should hoist them. They are all pretty simple.

The rule of thumb is to hoist everything you can. If it's used only once, then it will be inlined by the compiler.

File: Plutarch/Rational.hs [hunk 5]
```diff
@@ -1,28 +1,125 @@
...
+
+instance PEq PRational where
+  l #== r =
+    pmatch l $ \(PRational ln ld) ->
+      pmatch r $ \(PRational rn rd) ->
+        rd * ln #== rn * ld
+
+instance POrd PRational where
+  l #<= r =
+    pmatch l $ \(PRational ln ld) ->
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Rational.hs:33]: @TotallyNotChase this should probably also be in the guide

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Rational.hs:33]: #92

**Review by kozross — DISMISSED** [review]
Looks good to me. Are there plans to add `round`, `truncate` and/or `properFraction` support?

**Comment by L-as** [comment]
Needs to merge in upstream.

**Comment by L-as** [comment]
> Looks good to me. Are there plans to add `round`, `truncate` and/or `properFraction` support?

If you want to add them, sure.

**Review by L-as — DISMISSED** [review]
LGTM

**Comment by Geometer1729** [comment]
@srid I tried to write a PIsData instance but I'm not really sure if this is right.

**Comment by Geometer1729** [comment]
>

**Comment by Geometer1729** [comment]
I'm sorry I clicked the wrong button  and closed this.

**Comment by Geometer1729** [comment]
> Looks good to me. Are there plans to add `round`, `truncate` and/or `properFraction` support?

I kinda forgot about those. Looks like we do need them. I'll start working on that.

**Comment by srid** [comment]
> @srid I tried to write a PIsData instance but I'm not really sure if this is right.

You can add a test to `examples/Examples/Rationals.hs` to verify that `pfromData . pdata` is essentially `id`, on a bunch of inputs (property tests would be better, but we don't use them yet).

**Comment by Geometer1729** [comment]
@srid that test revealed a bug, I fixed it and now the tests pass, hopefully that was the only bug.
@kozross I implemented the rounding functions you mentioned.

File: Plutarch/Rational.hs [hunk 6]
```diff
@@ -1,3 +1,4 @@
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
```
Comment (L-as) [inline Plutarch/Rational.hs:1]: ?

File: Plutarch/Rational.hs [hunk 7]
```diff
@@ -11,12 +12,33 @@ module Plutarch.Rational (
...
 import Plutarch.Integer (PInteger, PIntegral (pdiv, pmod))
+import Plutarch.List (PListLike (pcons, phead, pnil, ptail), pmap)
 
 data PRational s = PRational (Term s PInteger) (Term s PInteger)
 
+instance PIsData PRational where
+  pfromData x = pListToRat #$ pmap # pasInt #$ pasList # pforgetData x
+  pdata x =
+    (punsafeCoerce :: Term _ (PAsData (PBuiltinList (PAsData PInteger))) -> Term _ (PAsData PRational)) $
+      pdata $ pmap # plam pdata #$ pRatToList # x
```
Comment (L-as) [inline Plutarch/Rational.hs]: Why not just do `pdata $` instead of `plam pdata #$`?

Comment (L-as) [reply Plutarch/Rational.hs]: Otherwise this looks fine actually :+1: , however, there is a design choice to be made here, between this approach and another approach. @TotallyNotChase

**Review by L-as — CHANGES_REQUESTED** [review]

File: Plutarch/Rational.hs (hunk 6 above)
Comment (L-as) [reply Plutarch/Rational.hs:1]: I don't think we need the other pragma either.

**Comment by L-as** [comment]
In general, I recommend not adding more features to PRs. You should just make a new PR.

File: Plutarch/Rational.hs (hunk 7 above)
Comment (L-as) [reply Plutarch/Rational.hs]: #56

Comment (L-as) [reply Plutarch/Rational.hs]: I just noticed these aren't hoisted.

File: Plutarch/Rational.hs (hunk 6 above)
Comment (kozross) [reply Plutarch/Rational.hs:1]: This is a _really_ bad thing to enable in general.

File: Plutarch/Rational.hs [hunk 8]
```diff
@@ -1,28 +1,235 @@
...
+pnumerator :: Term s (PRational :--> PInteger)
+pnumerator = phoistAcyclic $ plam $ \x -> pmatch x $ \(PRational n _) -> n
+
+pdenominator :: Term s (PRational :--> PInteger)
+pdenominator = phoistAcyclic $ plam $ \x -> pmatch x $ \(PRational _ d) -> d
+
+pfromInteger :: Term s (PInteger :--> PRational)
+pfromInteger = phoistAcyclic $ plam $ \n -> pcon $ PRational n 1
+
+pround :: Term s (PRational :--> PInteger)
```
Comment (kozross) [inline Plutarch/Rational.hs:218]: Does this match the behaviour of `round` from both Plutus and `base`? Specifically:

``round x`` returns the nearest integer to ``x``; the even integer if ``x`` is equidistant between two integers

File: examples/Examples/Rationals.hs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    , testCase "round 5/3" $
+        expect $ pround # (5 / 3) #== 2
+    , testCase "round 4/3" $
+        expect $ pround # (4 / 3) #== 1
+    , testCase "round 5/2" $
+        expect $ pround # (5 / 2) #== 3
+    , testCase "round 9/4" $
+        expect $ pround # (9 / 4) #== 2
+    , testCase "round 11/4" $
+        expect $ pround # (11 / 4) #== 3
```
Comment (kozross) [inline examples/Examples/Rationals.hs:39]: We need to cover the 'equidistant' case described above. There are also no tests dealing with negative values.

File: examples/Examples/Rationals.hs [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    , testCase "truncate 7/4" $
+        expect $ ptruncate # (7 / 4) #== 1
+    , testCase "properFraction 11/7" $
+        expect $
+          pmatch (pproperFraction # (11 / 7)) $ \(PPair x y) ->
+            x #== 1 #&& y #== (4 / 7)
+    , testCase "properFraction 13/7" $
+        expect $
+          pmatch (pproperFraction # (13 / 7)) $ \(PPair x y) ->
+            x #== 1 #&& y #== (6 / 7)
```
Comment (kozross) [inline examples/Examples/Rationals.hs:53]: Similarly here: we need to at least deal with negative values properly.

**Review by kozross — CHANGES_REQUESTED** [review]

File: Plutarch/Rational.hs (hunk 6 above)
Comment (Geometer1729) [reply Plutarch/Rational.hs:1]: Sorry about that, I think the haskell-language-server plugin I'm trying out added those. I'm not sure why. I've removed both pragmas.

File: Plutarch/Rational.hs (hunk 8 above)
Comment (Geometer1729) [reply Plutarch/Rational.hs:218]: It does not I didn't realize that was the behavior of round. I'll fix that.

**Comment by L-as** [comment]
Please retarget to staging BTW>

**Comment by Geometer1729** [comment]
> In general, I recommend not adding more features to PRs. You should just make a new PR.

Sorry about that, I'll keep that in mind going forward.

**Comment by Geometer1729** [comment]
I think I've made all of the requested changes now.

**Review by kozross — APPROVED** [review]

**Comment by L-as** [comment]
This has merge conflicts now unfortunately.

**Comment by L-as** [comment]
I think I solved it using GH's built-in functionality.

**Comment by L-as** [comment]
Formatting is wrong FWIW. `./bin/format` needs to be run. CI is broken because staging branch is broken.
